[OpenMP] Add tests and benchmarks

klaa97 opened the pull request:
# TAFFO x OpenMP

This PR is one of four PRs concerning the integration of TAFFO with OpenMP:
1. Initializer PR, patching the runtime OpenMP library functions with a trampoline call (https://github.com/HEAPLab/TAFFO-submodule-init/pull/4)
2. Conversion PR, re-creating the correct runtime OpenMP library functions from the trampoline calls, after the conversion in fixed points (https://github.com/HEAPLab/TAFFO-submodule-conv/pull/4)
3. Test PR, with tests and the integration of the [PolyBench-ACC benchmark](https://github.com/cavazos-lab/PolyBench-ACC) 
4. TAFFO high-level PR, with the technical documentation (https://github.com/HEAPLab/TAFFO/pull/4)


## Summary
This PR adds:
- Comprehensive tests on the supported OpenMP constructs
- The [PolyBench-ACC benchmark](https://github.com/cavazos-lab/PolyBench-ACC) adapted with TAFFO annotations, and heavily modified to make it actually working

klaa97 committed:
Experimental first OpenMP benchmark

klaa97 committed:
Tentative openmp benchmark pragmas fix

klaa97 committed:
Support floyd-warshall and (experimental) datamining benchmark for OMP

klaa97 committed:
Initial simple tests for OpenMP

klaa97 committed:
Add test cases for OMP critical and arrays

klaa97 committed:
Add minimal test cases for OMP reduction investigation

klaa97 committed:
Fix OpenMP reduction tests

klaa97 committed:
Fix floyd warshall benchmark (with double/float datatype)

klaa97 committed:
Adapt covariance benchmark annotations

klaa97 committed:
Delete correlation benchmark for design issues

klaa97 committed:
Fix covariance to convert in fixp

michiari committed:
Test case annotations fixed.

klaa97 committed:
Adapt 2mm and 3mm benchmark

klaa97 committed:
Add annotations to the remaining linear algebra kernels

klaa97 committed:
Add annotations to remaining benchmarks

klaa97 committed:
Add details about the deleted benchmarks in the README

klaa97 committed:
Add test for OpenMP tasks

klaa97 committed:
Enhance reduction test and add test for sections

klaa97 committed:
Update README for OpenMP benchmarks

skeru commented on polybench-c-openmp/README.md:
````diff
@@ -0,0 +1,66 @@
+
+First, do
+```bash
+$ export LLVM_DIR=/path/to/LLVM/install
````
can we re-write it by making use of `llvm-config`?

klaa97 replied:
I don't get what do you mean by making use of `llvm-config` - isn't `llvm-config`  useful only to check which llvm installation is being exported and will be used?

I copied the steps from the other README in here https://github.com/HEAPLab/TAFFO-test/tree/develop/polybench-c-4.2.1-beta

skeru replied:
I meant to use
```sh
$ export LLVM_DIR=`llvm-config --prefix`
```

skeru replied:
`llvm-config` can be used to print out several details related to the version of llvm being used, including the install prefix.
Custom installations of LLVM will have custom `llvm-config` executables.

klaa97 replied:
Ahh! Makes sense! Thanks for the clarification, going to fix 👌

danielecattaneo replied:
To be honest I don't know if I like using llvm-config because it forces the benchmark to use the LLVM install currently on the PATH (which might be release LLVM and not the debug LLVM I want to use), and additionally it introduces a discrepancy between the old polybench scripts and the OpenMP scripts.

I would ignore this suggestion for now, and postpone it to later internal evaluation between me, @skeru and @michiari

michiari replied:
I agree with @danielecattaneo on this one.

skeru replied:
In principle, you can parametrise the `llvm-config` you want to use and use that one everywhere in TAFFO.
However, I do agree on the fact we should uniform this setting all across the code base and be consistent with it.

We can move this discussion internally and have this PR being merged without this change.

skeru reviewed (commented):
LGTM

klaa97 committed:
Fix wrong annotation in test

danielecattaneo commented on polybench-c-openmp/README.md:
```diff
@@ -0,0 +1,66 @@
+
+A not-negligible part of the benchmarks has not only syntactical, but also logical errors, when compiling with plain clang. Therefore, they have been excluded and deleted.
+The deleted benchmarks, with the related motivation, are the following:
+- datamining/correlation - Segmentation Fault
+- linear-algebra/kernels/atax - Segmentation Fault
+- linear-algebra/kernels/bicg - Segmentation Fault
+- linear-algebra/kernels/cholesky - Segmentation Fault
+- linear-algebra/kernels/gemver - Segmentation Fault
+- linear-algebra/kernels/gesummv - Segmentation Fault
+- linear-algebra/kernels/mvt - Segmentation Fault
+- linear-algebra/kernels/trisolv - Segmentation Fault
+- linear-algebra/kernels/trmm - Segmentation Fault
+- linear-algebra/solvers/durbin - Segmentation Fault
+- linear-algebra/solvers/dynprog - Segmentation Fault
+- linear-algebra/solvers/gramschmidt - Segmentation Fault
+- linear-algebra/solvers/lu - The results of the benchmark are mostly NaN
+- linear-algebra/solvers/ludcmp - The results of the benchmark are mostly NaN
+- stencils/adi - Although the benchmark should output numbers between 0 and 1, the resulting numbers are extremely high
+- stencils/convolution-2d - Segmentation Fault
+- stencils/convolution-3d - Segmentation Fault
+- stencils/fdtd-apml - Segmentation Fault
```
Holy shit

danielecattaneo commented on polybench-c-openmp/datamining/covariance/Makefile:
```diff
@@ -0,0 +1,2 @@
+-include ../../utilities/options.mk
+-include ../../utilities/c.mk
```
Add a newline at the end of the file

danielecattaneo commented on simple-test-cases/openmp_array1.c:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    printf("result: %f\n", result);
+  }
```
Missing newline at the end of file

danielecattaneo commented on simple-test-cases/openmp_array_critical1.c:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  printf("result: %f\n", result);
+}
```
Missing newline again (not going to comment these any further, fix them all)

danielecattaneo commented on simple-test-cases/openmp_reduction1.c:
```diff
@@ -0,0 +1,25 @@
+
+  int i = 0;
+
+#pragma omp parallel for reduction(+:result)
```
```suggestion
  #pragma omp parallel for reduction(+:result)
```
It's more legible if you indent the openmp pragmas. Apply this everywhere.

danielecattaneo commented on simple-test-cases/openmp_section1.c:
```diff
@@ -0,0 +1,29 @@
+#define N (100)
+
+int main(int argc, char *argv[]) {
+  float foo __attribute((annotate("target('foo') scalar(range(0,10) final)"))) =
+      1.0f;
+  float pop
+      __attribute((annotate("target('pop') scalar(range(0,2000) final)"))) =
+          363;
+  float result
+      __attribute((annotate("target('result') scalar(range(0,2000)) final)"))) =
+          0;
```
These look ugly but I'm not sure if the alternatives are prettier or not. In doubt, leave them as they are.

klaa97 replied:
I tried to use better-named variables 👌 to make the test more understandable

danielecattaneo replied:
I was actually referring to the indentation... But whatever, it's not an important issue, especially in tests.

danielecattaneo commented on simple-test-cases/openmp_section1.c:
```diff
@@ -0,0 +1,29 @@
+      __attribute((annotate("target('result') scalar(range(0,2000)) final)"))) =
+          0;
+
+#pragma omp parallel num_threads(4)
+  {
```
Is it possible to have the open brace on the same line as the pragma?

klaa97 replied:
It should not be possible: the main reason is that the pragma is read from the pre-processor, so the extra tokens are ignored from the compiler

In particular, the following warning is issued:
`./openmp_task1.c:16:37: warning: extra tokens at the end of '#pragma omp parallel' are ignored [-Wextra-tokens]`

And therefore the open brace is ignored

danielecattaneo reviewed (changes requested):
Polybench part: it's polybench, what else there is to say? Slightly missed the opportunity to rewrite my old stinking scripts from scratch, but at least they are consistent with the other polybench versions in the repo, so that's a plus.

Simple tests: overall nice work, even though I expected more of them (not an issue though). Again, my only objections are purely cosmetic.

klaa97 committed:
Add new openmp tests and cleanup old tests

bebora committed:
Add trailing newline where missing

bebora committed:
Improve style

Indent OpenMP pragmas and add newlines before functions' brackets

bebora committed:
Add not working Polybench OpenMP benchmarks

They may be useful in the future while trying to understand why they
don't work.

bebora committed:
Restore original README

bebora committed:
Add trailing newline where missing

bebora committed:
Add diff between original and annotated OpenMP PolyBench

bebora committed:
Update modifications info in README.md

klaa97 committed:
Remove additional README

klaa97 committed:
Use STANDARD_DATASET to keep consistency with original benchmark

klaa97 committed:
Fix 2mm standard datasets value

danielecattaneo reviewed (approved):
Approved :)
